### ISSUE TITLE
Project references, builds project multiple times just for adding a space ?

### ISSUE THREAD

**Issue by appsolutegeek** [Microsoft/TypeScript#26545]
Adding a space to a file in pkg2 causes multiple builds of itself and pkg1 (pkg1 has pkg2 as a dependency)

I have setup the new project references. I have 2 simple pkg folders, pkg1 has a dependency on pkg2.

here is the pkg1 tsconfig

```
{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "target": "es2018",
    "skipLibCheck": true,
    "lib": [
      "esnext"
    ],
    "rootDir": "src",
    "outDir": "lib",
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "paths": {
      "@test/pkg2": ["./packages/pkg2/src"]
    }
  },
  "references": [
    { "path": "../pkg2" }
  ],
  "exclude": ["node_modules", "lib"]
}
```

and pkg2 (has composite set)

```
{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "composite":true,
    "target": "es2018",
    "skipLibCheck": true,
    "lib": [
      "esnext"
    ],
    "rootDir": "src",
    "outDir": "lib",
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
  },
  "exclude": ["node_modules", "lib"]
}
```

**Code**

```ts
tsc -b --verbose --watch
```

produces multiple build steps, seems pkg1 and pkg2 are built multiple times.

Here is a output for placing a space inside the index.ts file - found in pkg2.

```
message TS6350: Project '../pkg2/tsconfig.json' is out of date because oldest output '../pkg2/lib/index.js' is older than newest input '../pkg2/src/index.ts'

message TS6358: Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg2/tsconfig.json'...

message TS6350: Project 'tsconfig.json' is out of date because oldest output 'lib/index.js' is older than newest input 'src/index.ts'

message TS6358: Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg1/tsconfig.json'...

message TS6358: Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg2/tsconfig.json'...

message TS6358: Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg1/tsconfig.json'...

message TS6358: Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg2/tsconfig.json'...

message TS6358: Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg1/tsconfig.json'...

message TS6358: Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg2/tsconfig.json'...

message TS6358: Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg1/tsconfig.json'...

message TS6358: Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg2/tsconfig.json'...

message TS6358: Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg1/tsconfig.json'...

message TS6358: Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg1/tsconfig.json'...

message TS6358: Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg1/tsconfig.json'...

message TS6358: Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg1/tsconfig.json'...

message TS6358: Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg1/tsconfig.json'...

message TS6358: Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg1/tsconfig.json'...
```

This seems to be causing a chain of builds ?

**Comment by sheetalkamat** [comment]
Can you please try typescript@next generated tonight (that is anything that's equal to or above typescript@3.1.0-dev.20180829) to see if it fixes the issue. thanks

**Comment by appsolutegeek** [comment]
Hi, errmm, a little strange, the first time i change pkg2 then I get file change detected and then starts building the pkg2 and then pkg1....  once it finsihes it states "[3:48:20 PM] Found 0 errors. Watching for file changes." - No loop as far as I know... 

But in watch mode, it keeps clearing the screen, i tried passing --noEmit

Anyway, on subsequent changes, it states that file change detected and starting incremental compilation but doesn't output anything else.. i.e. No "[3:48:20 PM] Found 0 errors. Watching for file changes." i.e. it displays

```
[3:48:32 PM] File change detected. Starting incremental compilation...
```

Is this the expected behaviour? seems odd.

It seems it hasn't finished or certainly doesn't state

"Found 0 errors. Watching for file changes."

The first pass it went through the stages of saying, File Change detected, starting incremental build and then display Found 0 errors and waiting for changes..

**Comment by sheetalkamat** [comment]
> But in watch mode, it keeps clearing the screen, i tried passing --noEmit

you need to pass `--preserveWatchOutput` to not let screen clear. It remains same as out `--watch` behaviour when building without `--build`

> The first pass it went through the stages of saying, File Change detected, starting incremental build and then display Found 0 errors and waiting for changes..

Fixed that in #26742

**Comment by appsolutegeek** [comment]
thanks, its seems better now, just update the latest next.. Considering i have pkg1 that depends on pkg2

Making a change in pkg2 i get

```
[9:57:12 AM] Found 0 errors. Watching for file changes.

[9:57:20 AM] File change detected. Starting incremental compilation...

[9:57:20 AM] Project '../pkg2/tsconfig.json' is out of date because oldest output '../pkg2/lib/index.js' is older than newest input '../pkg2/src/index.ts'

[9:57:20 AM] Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg2/tsconfig.json'...

[9:57:21 AM] Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg1/tsconfig.json'...

[9:57:21 AM] Found 0 errors. Watching for file changes.
```

so looks ok BUT making a change in pkg1 i get multiplem "[10:02:45 AM] File change detected. Starting incremental compilation..."

```
[10:02:45 AM] File change detected. Starting incremental compilation...

[10:02:45 AM] File change detected. Starting incremental compilation...

[10:02:45 AM] File change detected. Starting incremental compilation...

[10:02:45 AM] File change detected. Starting incremental compilation...

[10:02:46 AM] Project 'tsconfig.json' is out of date because oldest output 'lib/index.js' is older than newest input 'src/index.ts'

[10:02:46 AM] Building project '/Users/ian/Development/Scratch/lerna-new-test/packages/pkg1/tsconfig.json'...

[10:02:46 AM] Found 0 errors. Watching for file changes.
```

Thanks in advance

**Comment by sheetalkamat** [comment]
@appsolutegeek The order of build, i dont see the issue you mentioned. Infact we have a test for this where we verify the order of build (https://github.com/Microsoft/TypeScript/blob/master/src/testRunner/unittests/tsbuildWatchMode.ts#L133) Do you mind sharing the code and change that repros this issue.

Fixed the multiple notification of  File change detected. Starting incremental compilation... and should be available in tonights drop

**Comment by appsolutegeek** [comment]
Just installed the @next and tried again, and this problem seems to be fixed. thanks. When we it be released, do you have an idea ? Thanks

**Comment by sheetalkamat** [comment]
This will be available in 3.1 when its released. Thanks
